### PR TITLE
WIP: Visualize multichain data

### PR DESCRIPTION
**Pull request by rkeulemans**
Solves #50 #51 #52

### CODE REVIEW

**Comment by rkeulemans** [comment]
Because the logic begins to mess up I have the following suggestion:
- Introduce a MVC pattern for the Trust Display;
- An observable Model of the network (that can possibly self-update 'real-time').
- A Controller that handles mouse clicks hovers etc. and registers as an observer to the Model.
- At least two View's, registering as observers to the Model:
  - A View containing the whole Trust Display as a QWidget
  - A View, embedded in the Trust Display page, containing the pyplot graph and the functions that change it, e.g. draw_node(), draw_node() etc.

File: TriblerGUI/widgets/graph_provider.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
 """
 Provides a class which encapsulates the graph construction logic.
 """
+from __future__ import division
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/graph_provider.py]: Please don't use this. This is very confusing when reading the code if one doesn't know this is being imported.

File: TriblerGUI/widgets/graph_provider.py [hunk 2]
```diff
@@ -1,6 +1,8 @@
...
+import math
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/graph_provider.py]: Only use import-from statements.

File: TriblerGUI/widgets/graph_provider.py [hunk 3]
```diff
@@ -10,25 +12,77 @@ class GraphProvider():
...
-        pos = [(2, 2), (1, 1), (2, 3), (3, 1)]
+        graph = nx.Graph()
+
+        display_information = network_model.retrieve_display_information()
+
+
+
+
+
+        """ Dummy data of a focus node and its peers """
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/graph_provider.py]: Use a single-line comment.

File: TriblerGUI/widgets/graph_provider.py [hunk 4]
```diff
@@ -10,25 +12,77 @@ class GraphProvider():
...
+        radius = 1
 
+        # Angle between neighbors
+        relative_angle = (2 * math.pi) / node_count
+
+        # Compute position of peers
+        self.peers = [polar_to_cartesian(center_x, center_y, radius, x * relative_angle) for x in range(node_count)]
+
+        def draw_node(pos, color):
+            """ Draw a networkx node with a specified color """
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/graph_provider.py]: Please start the contents of a multi-line comment on a new line and end sentences with a `.`.

File: TriblerGUI/widgets/network_model.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
+"""
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/network_model.py]: Could you place all GUI files in a new package?

File: TriblerGUI/widgets/network_model.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        self.neighbor_level = neighbor_level
+
+    def retrieve_display_information(self):
+        """
+        Retrieve the Trust Display information from the HTTP API.
+
+        By using the focus_node and neighbor_level attributes from the TrustDisplayPage object, query the display HTTP
+        endpoint to retrieve the information which can be used in the display.
+        """
+        url = "display?focus_node={0}".format(self.focus_node)
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/network_model.py]: Please use the `%` operator as it is the convention in Tribler.

File: TriblerGUI/widgets/network_model.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        :param response: the JSON dictionary passed on by the HTTP Request
+        """
+        focus_node = response["focus_node"]
+
+        grouped_edges = NetworkModel.group_elements(response["edges"], "from")
+
+        # TODO: Do the procedure below for each of the nodes in the graph, whilst erasing duplicates.
+        focus_node_edges = NetworkModel.get_combined_edges(grouped_edges, focus_node)
+
+        # TODO: Send the data to the display to display the data
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/network_model.py]: Do we want to send data to the display? Is this part of some observer pattern?

File: TriblerGUI/widgets/network_model.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        """
+        Combine the directed edges between the given node and other nodes to one edge per pair.
+
+        The attributes of the combined edges are calculated as follows:
+            - from: node_name
+            - to: to attribute from outgoing edge from node_name
+            - amount_up: amount from the outgoing edge from node_name
+            - amount_down: amount from the ingoing edge to node_name if any
+            - ratio: amount_up / (amount_up + amount_down)
+            - log_ratio: log(amount_up + 1) / (log(amount_up + 1) + log(amount_down + 1))
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/network_model.py]: It is unclear what the reason behind this `log_ratio` is.

File: TriblerGUI/widgets/network_model.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+            # Find the inverse edge by looking in the group of the "to" node in the original edge.
+            inverse_edge = next((inv for inv in grouped_edges[edge["to"]] if inv["to"] == node_name), None)
+            up = edge["amount"]
+            down = 0
+            ratio = 0
+            log_ratio = 0
+            if inverse_edge:
+                down = inverse_edge["amount"]
+            if not (up == 0 and down == 0):
+                ratio = float(up) / (up + down)
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/network_model.py]: Isn't this more like a fraction, rather than a ratio?

File: TriblerGUI/widgets/network_model.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+        Group elements in the list by the given attribute.
+
+        Use a defaultdict to group the elements by their attribute. This function returns a dictionary with keys
+        corresponding to the different occurring attributes.
+
+        :param elements: the list of elements which has to be grouped
+        :param attribute: the attribute from the elements which the elements must be grouped on
+        :return: a defaultdict with elements grouped by attribute name 
+        """
+        return_dict = defaultdict(list)
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/network_model.py]: `defaultdict`'s can be confusing and may lead to unexpected behavior. Avoid these if you can.

File: TriblerGUI/widgets/network_model.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        for element in elements:
+            return_dict[element[attribute]].append(element)
+
+        return return_dict
+
+    def get_focus_node(self):
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/network_model.py]: Please remove these getters and setters.

File: TriblerGUI/widgets/trustpage.py [hunk 12]
```diff
@@ -24,6 +25,12 @@ def initialize_trust_page(self):
         Load the pyplot graph into the QWidget.
         """
         vertical_layout = self.window().network_widget.layout()
+
+        network_model = NetworkModel(focus_node=u"30")
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage.py]: What is this `30`?

File: TriblerGUI/widgets/trustpage.py [hunk 13]
```diff
@@ -24,6 +25,12 @@ def initialize_trust_page(self):
...
         graph_data = GraphProvider()
-        self.network_graph = FigureCanvas(graph_data.provide_figure())
+        self.network_graph = FigureCanvas(graph_data.provide_figure(network_model))
+        self.network_graph.mpl_connect('button_press_event', graph_data.handle_mouseclick)
+
+
+
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage.py]: Too much newlines.

File: run_tribler.py [hunk 14]
```diff
@@ -1,6 +1,7 @@
 import os
 import sys
 import multiprocessing
+import logging  # TODO: Remove this import
```
Comment (vandenheuvel) [inline run_tribler.py]: If you don't like all this logging, maybe keep these edits and add it to your gitignore?

**Pull request closed without merging**